### PR TITLE
Create an mpirun dedicated executable

### PR DESCRIPTION
**Pull request by rhc54**
If we install the embedded versions or PRRTE and PMIx, then
OPAL_PREFIX needs to adjust the installed location for PMIx
as well as for PRRTE when launching via mpirun.

In mpirun: Detect OPAL_PREFIX and set the corresponding PRRTE and PMIx
prefixes. Take advantage of the simplified proxy detection
support in PRRTE

In MPI_Init: Detect OPAL_PREFIX and set the corresponding PMIx
prefix

Signed-off-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/2c1889d1372aaf2a225d2c308fb0311c

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/b6799d73cf67c5ace61bb3af042ec4a8

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/06a96a4b3fd9b2c70dfbd3e503e69f9e

**Comment by rhc54** [comment]
bot:aws:retest

**Comment by rhc54** [comment]
Quite odd. @bwbarrett is there someone over there who can take a closer look at your CI? It's segfaulting, but no other CI can replicate (including me). Seems like it is something fundamental, but I have no idea how to even begin trying to decipher it. Unclear if it is timing out (since you have a timeout generating a SIGSEGV) or actually segfaulting.

**Comment by bwbarrett** [comment]
It's a real error, although for reasons I don't totally understand, it's not entirely deterministic in that it doesn't happen under gdb (oddly).  Stack trace is:

```
Reading symbols from mpirun...
[New LWP 361073]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `mpirun --get-stack-traces --timeout 180 --hostfile /home/bbarrett/h -np 2 hostn'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f6aa44d5675 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) where
#0  0x00007f6aa44d5675 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f6aa43ec503 in strdup () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x000055fc8bf0a4fb in main (argc=9, argv=0x7ffc709bd6f8) at main.c:75
```

`opal_install_dirs` is only filled out when `opal_init_util()` has been called, so I think the fix is to call `opal_init_util() at the start of `mpirun`.

**Comment by rhc54** [comment]
Thanks Brian - I've seen that kind of behavior before as it all depends on how memory was initialized (or not). I'd rather not call `opal_init_util` as it does so many things (e.g., setting up the memory interceptor) beyond just fill in the `installdirs` object, so I opted to only open that framework given that `opal_init_util` did just a few unrelated things before doing so. Hopefully, that will prove adequate.

**Review by gpaulsen — COMMENTED** [review]
Are the pmix/prrte submodule updates NEEDED for this to work, or just convenient to do it in one PR?

**Comment by rhc54** [comment]
> Are the pmix/prrte submodule updates NEEDED for this to work, or just convenient to do it in one PR?

Not sure I grok the context of this question, but let me assume you are asking about how to bring it to OMPI v5. It will have to be done as a cherry-pick, then "reset" the conflict (since the pmix and prrte branches are different), and then update the pmix and prrte release branch submodules. I haven't updated the pmix/prrte release branches yet, so nothing to do right now.

**Comment by awlauria** [comment]
Would it be easier to split up the commits? IE - a commit each to update the prrte/pmix pointers, and a commit for ompi changes.

**Comment by bwbarrett** [comment]
I don't know why it would be easier to split the commits.

That said, I'm also not sure that we should need a submodule pointer change at this point.  We did when mpirun was linking against libprrte.so, but since that isn't happening in the new design, we probably don't have the dependency we once did?

**Comment by rhc54** [comment]
You at least need to update the PRRTE submodule as it has a corresponding change to allow specification of the proxy schizo component

File: ompi/tools/mpirun/main.c [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/***************************************************************************
```
Comment (bwbarrett) [inline ompi/tools/mpirun/main.c:1]: shouldn't the license be at the top, not the bottom?

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: ompi/tools/mpirun/main.c (hunk 1 above)
Comment (rhc54) [reply ompi/tools/mpirun/main.c:1]: I left it as it was in prior OMPI versions. IIRC, it had to do with a splash shield for debuggers

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**